### PR TITLE
New Method: Implementation of Barzilai-Borwein Method

### PR DESCRIPTION
**Pull request by vp314**
# Overview

This PR implements both step-size strategies for the Barzilai-Borwein (BB) method as introduced in [this work](https://doi.org/10.1093/imanum/8.1.141). This PR closes issue #6

# Changes
- A data structure that contains the necessary information to implement the BB method 
- A constructor for the above data structure
- Logic for iterating the BB method
- Docstrings for all `struct`s and functions
- An example for using the BB method on the Gaussian Least Squares problem. 
- Tests for the BB method. Uses the Gaussian Least Squares method for the tests.

### CODE REVIEW

File: src/methods/gd_barzilai_borwein.jl [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+- `init_stepsize::T`, initial step size to start the method. 
+- `long_stepsize::Bool`, flag for step size; if true, use the long version of 
+    Barzilai-Borwein. If false, use the short version. 
+- `threshold::T`, gradient threshold. If the norm gradient is below this, then 
+    iteration stops.
+- `max_iterations::Int64`, max number of iterations (gradient steps) taken by 
+    the solver.
+- `iter_diff::Vector{T}`, a buffer for storing differences between subsequent
+    iterate values that are used for computing the step size
+- `grad_diff::Vector{T}`, a buffer for storign differences between gradient
```
Comment (VChristian) [inline src/methods/gd_barzilai_borwein.jl:21]: storing

File: src/methods/gd_barzilai_borwein.jl [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+If ``k=0``, then ``\\alpha_0`` is set to `optData.init_stepsize`. For ``k>0``,
+
+```math
+\\alpha_k = \\frac{(x_k - x_{k-1}^\\intercal (\\nabla f(x_k) - 
+    \\nabla f(x_{k-1})))}{\\Vert \\nabla f(x_k) - \\nabla f(x_{k-1})\\Vert_2^2}.
+```
+
+# Arguments
+
+- `optData::FixedStepGD{T}`, the specification for the optimization method.
```
Comment (VChristian) [inline src/methods/gd_barzilai_borwein.jl:155]: `optData::BarzilaiBorweinGD{T}`

File: test/methods/gd_barzilai_borwein.jl [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+        # Output
+        xk = barzilai_borwein_gd(optData, progData)
+        
+        # Stop Iteration should be 26
+        if optData.stop_iteration < optData.max_iterations 
+            @test optData.grad_val_hist[optData.stop_iteration] >= 
+                optData.threshold
+            @test optData.grad_val_hist[optData.stop_iteration + 1] < 
+                optData.threshold 
+        end
```
Comment (VChristian) [inline test/methods/gd_barzilai_borwein.jl:127]: Should also test if the correct step size is being taken.

Comment (VChristian) [reply test/methods/gd_barzilai_borwein.jl:127]: And if everything is being saved.

File: test/methods/gd_barzilai_borwein.jl [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+        bb_method = BarzilaiBorweinGD(real_type,
+            x0 = zeros(real_type, param_dim),
+            init_stepsize = real_type(1),
+            long_stepsize = true,
+            threshold = real_type(100.0),
+            max_iterations = 100)
+        
+        for (field_name, field_type) in zip(names, field_types)
+            @test typeof(getfield(bb_method, field_name)) == field_type
+        end
```
Comment (VChristian) [inline test/methods/gd_barzilai_borwein.jl:58]: Test the values of the fields

File: test/methods/gd_barzilai_borwein.jl [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+    end
+
+    ## test assertions
+    @test_throws AssertionError BarzilaiBorweinGD(Float64, 
+        x0=randn(10), 
+        init_stepsize = -0.3, #Assertion error, must be positive
+        long_stepsize = true,
```
Comment (VChristian) [inline test/methods/gd_barzilai_borwein.jl:65]: should also test `long_stepsize = false`

File: test/methods/gd_barzilai_borwein.jl [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+
+    ##########################################
+    # Test struct constructor
+    ##########################################
+
+    ## test types
+    real_types = [Int16, Int32, Int64, Float16, Float32, Float64]
+    param_dim = 10
+
+    for real_type in real_types
```
Comment (VChristian) [inline test/methods/gd_barzilai_borwein.jl:43]: Maybe also `for bool in [False, True]`

File: test/methods/gd_barzilai_borwein.jl (hunk 4 above)
Comment (vp314) [reply test/methods/gd_barzilai_borwein.jl:58]: I think the goal of this set of tests is just to make sure they are of the correct type.

**Review by VChristian — APPROVED** [review]

**Pull request merged**